waddrmgr/db.go

```diff
@@ -1865,3 +1816,165 @@ func upgradeToVersion3(namespace walletdb.Namespace, seed, privPassPhrase, pubPa
 	}
 	return nil
 }
+
+// upgradeToVersion4 upgrades the database from version 3 to version 4
+// The following fields were added to the serialized address row:
+// * 'watchingOnly' - boolean indicating whether the address is watching-only
+// This is required to flag each imported address as watching-only
+// * 'encryptedPubKeyHash' - encrypted pubkey hash of imported p2pkh address
+// This is required to import addresses without the associated public key.
+func upgradeToVersion4(namespace walletdb.Namespace) error {
+	err := namespace.Update(func(tx walletdb.Tx) error {
+		currentMgrVersion := uint32(4)
+		bucket := tx.RootBucket().Bucket(addrBucketName)
+
+		err := bucket.ForEach(func(k, v []byte) error {
+			// Skip buckets.
+			if v == nil {
+				return nil
+			}
+
+			// Insert watchingOnly bool at byte 14
+			// Since importing watching-only addrs was added in this version we
+			// can safely initialize this to falseByte
+			x := make([]byte, len(v)+1)
+			copy(x[:13], v[:13])
+			x[14] = falseByte
+			copy(x[15:], v[14:])
+
+			addrType := addressType(x[0])
+
+			// Update imported private key addresses for encryptedPubKeyHash field
+			switch addrType {
+			case adtImport:
+				// Insert nil for encrypted pkhash in rawData
+				rdlen := binary.LittleEndian.Uint32(x[15:19])
+				binary.LittleEndian.PutUint32(x[15:19], uint32(rdlen+4))
+				rawData := make([]byte, rdlen+4)
+				copy(rawData, x[19:])
+				binary.LittleEndian.PutUint32(x[19:23], 0)
+				copy(x[23:], rawData)
```

Comment:
The will cut off the end of the raw data.  `x` here will only be `len(v)+1` which doesn't include the extra 4 bytes, so when `rawData` is copied into `x[23:]` it will end up being 4 bytes too long.